wallet/src/wallet/mod.rs

```diff
@@ -187,19 +223,32 @@ impl<B: storage::Backend> Wallet<B> {
         self.accounts.keys()
     }
 
-    // TODO: this should not be public, as specified by BIP44 accounts must be created sequentially
-    // and a new next account should be rejected if the previously created one has no transactions
-    // associated with it
-    pub fn create_account(&mut self, account_index: U31) -> WalletResult<()> {
-        ensure!(
-            !self.accounts.contains_key(&account_index),
-            WalletError::AccountAlreadyExists(account_index)
-        );
+    pub fn number_of_accounts(&self) -> usize {
+        self.accounts.len() + self.unsynced_account.is_some() as usize
+    }
+
+    pub fn create_account(&mut self) -> WalletResult<U31> {
+        if self.unsynced_account.is_some() {
+            return Err(WalletError::LastAccountNotInSync);
+        }
+
+        let next_account_index = self.accounts.last_key_value().map_or(
+            Ok(U31::ZERO),
+            |(last_account_index, last_account)| {
+                if last_account.has_transactions() {
+                    last_account_index.plus_one().map_err(|_| WalletError::MaxNumAccountsExceeded)
+                } else {
+                    // Cannot create a new account if the latest created one has no transactions
+                    // associated with it
+                    Err(WalletError::EmptyLastAccount)
```

Comment:
Why?

Reply:
It is discouraged by bip44, if you create multiple consecutive empty accounts, the default algorithm for recovering a wallet's accounts will stop on the first empty account and not discover the rest.

Reply:
I see

Reply:
Should there also be a check that the transaction in the previous account has some sufficient number of confirmations? I can imagine a user who has a fresh transaction in account 0 creating account 1 immediately. The transaction in account 0 then gets reorged out. When re-syncing the wallet, whatever transactions may have been submitted to account 1 will not be discovered. May be left to a future discussion.

Reply:
Checking whether there are sufficient confirmations should also be centrally controlled somehow. The wallet should have a setting to view only confirmed transactions after some confirmations. But I view this as a nicity more than a necessity.